### PR TITLE
.eh_frame: Add debugging helpers

### PR DESCRIPTION
**Pull request by javierhonduco**
This PR introduces two changes:
- Removal of `expression_constants.go`, which we don't use
- Adding a new program `dist/eh-frame` to test and debug this stack unwinding method 

For the latter commit, some private methods and struct fields have been made public. These changes don't focus on API design or performance, we can undo this or redesign these interfaces to enable better debugging without exposing internal methods once the implementation works

## Test plan

```
[javierhonduco@fedora parca-agent]$ make && dist/eh-frame --executable ../test-programs/basic.o --filter-nops
=> Frame start: 401020, Frame end: 401046
instruction: DW_CFA_advance_loc
instruction: DW_CFA_undefined
=> Frame start: 401050, Frame end: 401055
=> Frame start: 401110, Frame end: 40111b
instruction: DW_CFA_advance_loc
instruction: DW_CFA_def_cfa_offset
instruction: DW_CFA_offset
instruction: DW_CFA_advance_loc
instruction: DW_CFA_def_cfa_register
instruction: DW_CFA_advance_loc
instruction: DW_CFA_def_cfa
=> Frame start: 401120, Frame end: 40112b
instruction: DW_CFA_advance_loc
instruction: DW_CFA_def_cfa_offset
instruction: DW_CFA_offset
instruction: DW_CFA_advance_loc
instruction: DW_CFA_def_cfa_register
instruction: DW_CFA_advance_loc
instruction: DW_CFA_def_cfa
=> Frame start: 401130, Frame end: 40114a
instruction: DW_CFA_advance_loc
instruction: DW_CFA_def_cfa_offset
instruction: DW_CFA_offset
instruction: DW_CFA_advance_loc
instruction: DW_CFA_def_cfa_register
instruction: DW_CFA_advance_loc
instruction: DW_CFA_def_cfa
```

### CODE REVIEW

File: pkg/stack/unwind/plan_table.go [hunk 1]
```diff
@@ -110,21 +112,21 @@ func (ptb *PlanTableBuilder) PlanTableForPid(pid int) (PlanTable, error) {
...
+		res = append(res, BuildTable(fdes))
 	}
 
 	// TODO(kakkoyun): Merge and order instructions of PlanTables.
 	// TODO(kakkoyun): Figure out mapping start address usage.
 	return res[0], nil
 }
 
-func (ptb *PlanTableBuilder) readFDEs(path string, start uint64) (frame.FrameDescriptionEntries, error) {
+func (ptb *PlanTableBuilder) ReadFDEs(path string, start uint64) (frame.FrameDescriptionEntries, error) {
```
Comment (kakkoyun) [inline pkg/stack/unwind/plan_table.go:129]: Let's not export `ReadFDEs` and `BuildTable`.
We can add a `PrintTable` function to the package to the same thing.

**Review by kakkoyun — COMMENTED** [review]
A couple of things I have pointed out.

File: internal/dwarf/frame/table.go [hunk 2]
```diff
@@ -19,7 +19,7 @@ type DWRule struct {
 
 // FrameContext wrapper of FDE context.
 type FrameContext struct {
-	loc           uint64
+	Loc           uint64
```
[the remark this answers is not in the document]
Comment (javierhonduco) [reply internal/dwarf/frame/table.go:22]: Good catch, was accessing it before and forgot to clean it up!

File: cmd/eh-frame/main.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	ptb := unwind.NewPlanTableBuilder(logger, process.NewMappingFileCache(logger))
+	fdes, err := ptb.ReadFDEs(executablePath, 0)
+	if err != nil {
+		// nolint
+		fmt.Println("failed with:", err)
+	}
+
+	for _, row := range unwind.BuildTable(fdes) {
+		// nolint
+		fmt.Printf("=> Frame start: %x, Frame end: %x\n", row.Begin, row.End)
```
[the remark this answers is not in the document]
Comment (javierhonduco) [reply cmd/eh-frame/main.go:59]: I was doing this initially, but I find the extra output (log level, etc) confusing when comparing addresses between tools. If it's ok, could we keep it as-is and use the logger later on?

[the remark this answers is not in the document]
Comment (kakkoyun) [reply cmd/eh-frame/main.go:59]: 👍 It's totally fine not to use it if this is the intended behaviour. You can disable the specific linter for the file or the package.

File: pkg/stack/unwind/plan_table.go [hunk 4]
```diff
@@ -116,14 +119,34 @@ func (ptb *PlanTableBuilder) PlanTableForPid(pid int) (PlanTable, error) {
...
 	}
 
 	// TODO(kakkoyun): Merge and order instructions of PlanTables.
 	// TODO(kakkoyun): Figure out mapping start address usage.
 	return res[0], nil
 }
 
+// PrintTable is a debugging helper that prints the unwinding table to stdout.
+// TODO(javierhonduco): accept a writer instead of directly printing to stdout.
+func (ptb *PlanTableBuilder) PrintTable(writer io.Writer, path string, start uint64, filterNops bool) error {
```
Comment (kakkoyun) [inline pkg/stack/unwind/plan_table.go:132]: Do we need the `start` as a parameter?

File: pkg/stack/unwind/plan_table.go [hunk 5]
```diff
@@ -170,7 +193,7 @@ func (ptb *PlanTableBuilder) readFDEs(path string, start uint64) (frame.FrameDes
 	return fde, nil
 }
 
-func buildTable(fdes frame.FrameDescriptionEntries) PlanTable {
+func BuildTable(fdes frame.FrameDescriptionEntries) PlanTable {
```
Comment (kakkoyun) [inline pkg/stack/unwind/plan_table.go:196]:
```suggestion
func buildTable(fdes frame.FrameDescriptionEntries) PlanTable {
```

File: pkg/stack/unwind/plan_table.go (hunk 4 above)
Comment (javierhonduco) [reply pkg/stack/unwind/plan_table.go:132]: Right now no, usually I wanna print the whole thing, thought we could add it as a parameter if necessary, but will remove it in the meantime

File: pkg/stack/unwind/plan_table_test.go [hunk 6]
```diff
@@ -29,7 +29,7 @@ func TestBuildPlanTable(t *testing.T) {
 	fdes, err := ptb.readFDEs("testdata/pie-dynamic", 0)
 	require.NoError(t, err)
 
-	planTable := buildTable(fdes)
+	planTable := BuildTable(fdes)
```
Comment (kakkoyun) [inline pkg/stack/unwind/plan_table_test.go:32]:
```suggestion
	planTable := buildTable(fdes)
```

File: internal/dwarf/frame/table.go [hunk 7]
```diff
@@ -321,7 +369,7 @@ func setloc(frame *FrameContext) {
 	var loc uint64
 	_ = binary.Read(frame.buf, frame.order, &loc)
 
-	frame.loc = loc + frame.cie.staticBase
+	frame.loc = loc
```
Comment (kakkoyun) [inline internal/dwarf/frame/table.go:372]: I think this change requires a comment. We might be missing to pass the correct base to the parser. This changes the initial ergonomics of the package.

Comment (kakkoyun) [reply internal/dwarf/frame/table.go:372]: The `staticBase` is passed to the parser correctly. What did go wrong here?

Comment (javierhonduco) [reply internal/dwarf/frame/table.go:372]: My bad, left some testing code here

File: pkg/stack/unwind/plan_table_test.go (hunk 6 above)
Comment (javierhonduco) [reply pkg/stack/unwind/plan_table_test.go:32]: Oops, should have run the tests

**Review by kakkoyun — APPROVED** [review]

**Pull request merged**